pkg/operator/etcdcertsigner/etcdcertsignercontroller.go

```diff
@@ -331,6 +341,81 @@ func (c *EtcdCertSignerController) createSecretForNode(node *corev1.Node) error
 	return nil
 }
 
+// Get43CompatibilityNames checks for the existance of FQDN used for etcd peerURLs. If found this means
+// the SAN for peer should contain either an explicit DNS record IE etcd-1.cluster.com or fall back
+// to wildcard if no 4.3 etcd naming is observed we return nil
+func (c *EtcdCertSignerController) Get43CompatibilityNames(node *corev1.Node) []string {
+	var errs []string
+	etcdDiscoveryDomain, err := c.getEtcdDiscoveryDomain()
+	if err != nil {
+		errs = append(errs, err.Error())
+	}
+	nodeInternalIPs, err := dnshelpers.GetInternalIPAddressesForNodeName(node)
+	if err != nil {
+		errs = append(errs, err.Error())
+	}
+	etcdMembers, err := c.etcdClient.MemberList()
+	if err != nil {
+		errs = append(errs, err.Error())
+	}
+	peerHostNames, err := getPeerDNSSubjectAlternativeNamesFromMemberList(etcdDiscoveryDomain, etcdMembers, nodeInternalIPs)
+	if err != nil {
+		errs = append(errs, err.Error())
+	}
+
+	// fallback to wildcard for 4.3 compatability if error is observed
+	if len(errs) > 0 {
+		return []string{"*." + etcdDiscoveryDomain}
+	}
+	return peerHostNames
+}
+
+func getPeerDNSSubjectAlternativeNamesFromMemberList(etcdDiscoveryDomain string, etcdMembers []*etcdserverpb.Member, nodeInternalIPs []string) ([]string, error) {
+	var errs []string
+	var found bool
+
+	hosts := []string{}
+	for _, m := range etcdMembers {
+		if len(m.PeerURLs) > 0 && len(m.ClientURLs) > 0 {
+			// 4.3 naming
+			if strings.HasPrefix(m.Name, "etcd-member") {
+				found = true
+			}
+			peer, err := url.Parse(m.PeerURLs[0])
+			if err != nil {
+				errs = append(errs, err.Error())
+				continue
+			}
+			if addr := net.ParseIP(peer.Hostname()); addr != nil {
+				continue
+			}
+			client, err := url.Parse(m.ClientURLs[0])
+			if err != nil {
+				errs = append(errs, err.Error())
+				continue
+			}
+			for _, nodeIP := range nodeInternalIPs {
+				if nodeIP == client.Hostname() {
+					hosts = append(hosts, peer.Hostname())
+				}
+			}
+		}
+	}
+
+	switch {
+	case len(errs) > 0:
+		break // fallback to wildcard 4.3 compatability
+	case len(hosts) == 0 && found:
+		break // fallback to wildcard 4.3 compatability
+	case len(hosts) > 0:
+		return hosts, nil // 4.3 peer DNS found populate san with explicit records
+	case !found:
+		return nil, nil // 4.4
+	}
+
+	return hosts, fmt.Errorf("no peer SAN hostnames found in MemberList")
```

Comment:
it's weird to return a value and an error. and the value returned here is ignored.

Reply:
yeah.... fixed